In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sn

**Question:** load the data set in colab

In [4]:
from google.colab import files
uploaded = files.upload()  # Opens a file picker to upload file

Saving loan_data.csv to loan_data.csv


In [14]:
import io

for loan_data in uploaded.keys():
    print(f"Uploaded file: {loan_data}")
    df = pd.read_csv(io.BytesIO(uploaded[loan_data]))
    print(df.head())

Uploaded file: loan_data.csv
   person_age person_gender person_education  person_income  person_emp_exp  \
0        22.0        female           Master        71948.0               0   
1        21.0        female      High School        12282.0               0   
2        25.0        female      High School        12438.0               3   
3        23.0        female         Bachelor        79753.0               0   
4        24.0          male           Master        66135.0               1   

  person_home_ownership  loan_amnt loan_intent  loan_int_rate  \
0                  RENT    35000.0    PERSONAL          16.02   
1                   OWN     1000.0   EDUCATION          11.14   
2              MORTGAGE     5500.0     MEDICAL          12.87   
3                  RENT    35000.0     MEDICAL          15.23   
4                  RENT    35000.0     MEDICAL          14.27   

   loan_percent_income  cb_person_cred_hist_length  credit_score  \
0                 0.49               

**Question:** find the shape of dataset

In [6]:
df.shape

(45000, 14)

**Question**:The unique values of the output variable.


In [15]:
u=df.columns
for i in u:
  print(f"{df[i]},has unique value ")
  print(f"{df[i].unique()}")

0        22.0
1        21.0
2        25.0
3        23.0
4        24.0
         ... 
44995    27.0
44996    37.0
44997    33.0
44998    29.0
44999    24.0
Name: person_age, Length: 45000, dtype: float64,has unique value 
[ 22.  21.  25.  23.  24.  26. 144. 123.  20.  32.  34.  29.  33.  28.
  35.  31.  27.  30.  36.  40.  50.  45.  37.  39.  44.  43.  41.  46.
  38.  47.  42.  48.  49.  58.  65.  51.  53.  66.  61.  54.  57.  59.
  62.  60.  55.  52.  64.  70.  78.  69.  56.  73.  63.  94.  80.  84.
  76.  67. 116. 109.]
0        female
1        female
2        female
3        female
4          male
          ...  
44995      male
44996    female
44997      male
44998      male
44999      male
Name: person_gender, Length: 45000, dtype: object,has unique value 
['female' 'male']
0             Master
1        High School
2        High School
3           Bachelor
4             Master
            ...     
44995      Associate
44996      Associate
44997      Associate
44998       Bachelor
44

**Question**:A short paragraph in text form describing the dataset (included in the Colab notebook).


In [8]:
df.info()
df.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45000 entries, 0 to 44999
Data columns (total 14 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   person_age                      45000 non-null  float64
 1   person_gender                   45000 non-null  object 
 2   person_education                45000 non-null  object 
 3   person_income                   45000 non-null  float64
 4   person_emp_exp                  45000 non-null  int64  
 5   person_home_ownership           45000 non-null  object 
 6   loan_amnt                       45000 non-null  float64
 7   loan_intent                     45000 non-null  object 
 8   loan_int_rate                   45000 non-null  float64
 9   loan_percent_income             45000 non-null  float64
 10  cb_person_cred_hist_length      45000 non-null  float64
 11  credit_score                    45000 non-null  int64  
 12  previous_loan_defaults_on_file  

person_age person_gender person_education  person_income  person_emp_exp  \
0        22.0        female           Master        71948.0               0   
1        21.0        female      High School        12282.0               0   
2        25.0        female      High School        12438.0               3   
3        23.0        female         Bachelor        79753.0               0   
4        24.0          male           Master        66135.0               1   

  person_home_ownership  loan_amnt loan_intent  loan_int_rate  \
0                  RENT    35000.0    PERSONAL          16.02   
1                   OWN     1000.0   EDUCATION          11.14   
2              MORTGAGE     5500.0     MEDICAL          12.87   
3                  RENT    35000.0     MEDICAL          15.23   
4                  RENT    35000.0     MEDICAL          14.27   

   loan_percent_income  cb_person_cred_hist_length  credit_score  \
0                 0.49                         3.0           561   
1                 0.08                         2.0           504   
2                 0.44                         3.0           635   
3                 0.44                         2.0           675   
4                 0.53                         4.0           586   

  previous_loan_defaults_on_file  loan_status  
0                             No            1  
1                            Yes            0  
2                             No            1  
3                             No            1  
4                             No            1

**Ans**: There is a total of 45000 entries
the dataset has 14 columns named as above ,all the columns are non null all
there are 6+3 numerical data and 5 string data named as above.The final out put is whether the loan is approved or not contained the column named loan status
 **bold text**

**Question:** Preprocess the data:

*   Drop irrelevant columns.
*  Handle NaN values appropriately.



In [17]:
df.dropna(inplace=True) # dropping the null value
df=df.drop(columns=['loan_intent','previous_loan_defaults_on_file','person_gender','person_education','person_home_ownership'])
df.head()

person_age  person_income  person_emp_exp  loan_amnt  loan_int_rate  \
0        22.0        71948.0               0    35000.0          16.02   
1        21.0        12282.0               0     1000.0          11.14   
2        25.0        12438.0               3     5500.0          12.87   
3        23.0        79753.0               0    35000.0          15.23   
4        24.0        66135.0               1    35000.0          14.27   

   loan_percent_income  cb_person_cred_hist_length  credit_score  loan_status  
0                 0.49                         3.0           561            1  
1                 0.08                         2.0           504            0  
2                 0.44                         3.0           635            1  
3                 0.44                         2.0           675            1  
4                 0.53                         4.0           586            1

**Question**:  Split the dataset into training and testing sets.
* Create two models:

  *   One using the imported Logistic Regression model.
  *   Another by implementing a Logistic Regression model as outlined in the provided resources.





In [10]:
# Importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from google.colab import drive
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import MinMaxScaler

In [18]:
X=df.drop(columns=['loan_status'])
Y=df['loan_status']
Y.value_counts()

loan_status
0    35000
1    10000
Name: count, dtype: int64

In [19]:
scaler = MinMaxScaler()
X = scaler.fit_transform(X)


In [20]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.15,random_state=48)



In [21]:
model1=LogisticRegression()
model1.fit(X_train,Y_train)


LogisticRegression()

In [22]:
Ypred1=model1.predict(X_test)
accuracy1=accuracy_score(Y_test,Ypred1)*100


In [23]:
class LogitRegression():
  def __init__(self,learning_rate,iterations):

    self.learning_rate = learning_rate
    self.iterations = iterations

  def fit(self,X,Y):

    self.m, self.n = X.shape
    self.W = np.zeros(self.n)
    self.b = 0
    self.X = X
    self.Y = Y

    for i in range(self.iterations):
      self.update_weights()

    return self

  def update_weights(self):

    A = 1 / (1+ np.exp (- ( self.X.dot( self.W ) + self.b )))
    tmp = A - self.Y.T
    tmp = np.reshape( tmp, self.m )
    dW = np.dot( self.X.T, tmp/self.m )
    db = np.sum(tmp)/self.m

    self.W = self.W-self.learning_rate *dW
    self.b = self.b-self.learning_rate *db

    return self

  def predict(self,X):

    Z = 1 / ( 1 + np.exp( -( X.dot( self.W ) + self.b )))
    Y = np.where( Z>0.5, 1, 0 )

    return Y


In [24]:
model2 = LogitRegression(learning_rate=0.01, iterations=1000)
model2.fit(X_train, Y_train)


In [28]:
Ypred2=model2.predict(X_test)
correctly_classified=0
total_test_count=0
count=0
for count in range(np.size(Y_test)):
  if Ypred2[count]==Y_test.iloc[count]:
    correctly_classified+=1
  total_test_count+=1
  count+=1
accuracy2=(correctly_classified/total_test_count)*100
print("Accuracy given by sklearn model :",accuracy1)
print("Accuracy given by model made from scratch:",accuracy2)


Accuracy given by sklearn model : 82.34074074074074
Accuracy given by model made from scratch: 77.77777777777779


* Accuracy given by sklearn model : 82.34074074074074
* Accuracy given by model made from scratch: 77.77777777777779

The sklearn model is more accurate than the model made from the scrath
